### PR TITLE
first version of solution

### PR DESCRIPTION
**Pull request by Sashk52**

### CODE REVIEW

File: src/main/java/com/dev/movietheatre/dao/impl/ShoppingCartDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    public ShoppingCart getByUser(User user) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<ShoppingCart> getByUserQuery = session.createQuery(
+                    "from ShoppingCart sc "
+                            + "left join fetch sc.tickets "
+                            + "where sc.user = :user", ShoppingCart.class);
+            getByUserQuery.setParameter("user", user);
+            return getByUserQuery.getSingleResult();
```
Comment (MykolaFatkullin) [inline src/main/java/com/dev/movietheatre/dao/impl/ShoppingCartDaoImpl.java]: You can simplify it:
```suggestion
            return session.createQuery(
                    "from ShoppingCart sc "
                            + "left join fetch sc.tickets "
                            + "where sc.user = :user", ShoppingCart.class)
                            .setParameter("user", user).getSingleResult();
```

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public void addSession(MovieSession movieSession, User user) {
+        Ticket ticket = new Ticket();
+        ticket.setMovieSession(movieSession);
+        ticket.setUser(user);
+        ticketDao.add(ticket);
+        ShoppingCart shoppingCartByUser = shoppingCartDao.getByUser(user);
+        List<Ticket> tickets = shoppingCartByUser.getTickets();
+        tickets.add(ticket);
+        shoppingCartDao.update(shoppingCartByUser);
```
Comment (MykolaFatkullin) [inline src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: You add the ticket to the List, but I don't sure what tickets are updating in the shopping cart. And you can simplify your code: 
```suggestion
        shoppingCartByUser.getTickets().add(ticket);
        shoppingCartDao.update(shoppingCartByUser);
```

**Review by MykolaFatkullin — APPROVED** [review]

File: src/main/java/com/dev/movietheatre/dao/impl/ShoppingCartDaoImpl.java (hunk 1 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/dao/impl/ShoppingCartDaoImpl.java]: thanks, i have not known this construction. done

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java (hunk 2 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/Main.java [hunk 3]
```diff
@@ -50,5 +52,16 @@ public static void main(String[] args) throws AuthenticationException {
         User loginedUser = authenticationService.login("Billy@ukr.net", "111");
         System.out.println("Registered user " + registeredUser.toString());
         System.out.println("Logined user " + loginedUser.toString());
+
+        ShoppingCartService shoppingCartService = (ShoppingCartService) injector
+                .getInstance(ShoppingCartService.class);
+        User registeredUser1 = authenticationService.register("Sam@ukr.net", "000");
+        shoppingCartService.addSession(movieSession,registeredUser1);
+        ShoppingCart shoopingCardbyUser1 = shoppingCartService.getByUser(registeredUser1);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]:
```suggestion
        ShoppingCart shoppingCartByUser1 = shoppingCartService.getByUser(registeredUser1);
```

File: src/main/java/com/dev/movietheatre/model/MovieSession.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
-    @ManyToOne (fetch = FetchType.LAZY)
+    @ManyToOne
     @JoinColumn(name = "movie_id")
     private Movie movie;
     @ManyToOne (fetch = FetchType.LAZY)
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java]: what about this?

File: src/main/java/com/dev/movietheatre/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.MapsId;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "shopping_card")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/ShoppingCart.java]:
```suggestion
@Table(name = "shopping_cart")
```

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    private ShoppingCartDao shoppingCartDao;
+    @Inject
+    private TicketDao ticketDao;
+
+    @Override
+    public void addSession(MovieSession movieSession, User user) {
+        Ticket ticket = new Ticket();
+        ticket.setMovieSession(movieSession);
+        ticket.setUser(user);
+        ticketDao.add(ticket);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: if there is no shopping cart for the user - we still save ticket to the db, is it ok?

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    public void registerNewShoppingCart(User user) {
+        ShoppingCart newShoppingCart = new ShoppingCart();
+        newShoppingCart.setUser(user);
+        shoppingCartDao.add(newShoppingCart);
+    }
+
+    @Override
+    public void clear(ShoppingCart shoppingCart) {
+        User user = shoppingCart.getUser();
+        ShoppingCart shoppingCardByUser = getByUser(user);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: do you need this?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/movietheatre/Main.java (hunk 3 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: done

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 4 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: this fetch lefted from last hw, it's not needed anymore, but it created LazyIntitializationException in my MovieSession for Movie. @ManyToOne and Fetch.EAGER default is more logical.  so i deleted Fetch.LAZY. Is't problem?

File: src/main/java/com/dev/movietheatre/model/ShoppingCart.java (hunk 5 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/ShoppingCart.java]: done

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java (hunk 7 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: no, i don't. deleted. done

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java (hunk 6 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: I don't think so. When we register user, we register new shopping cart for him (Hibernate-06 Checklist. 5.Use method registerNewShoppingCart() in AuthenticationServiceImpl class in register() method.). So, we save ticket to existed shopping card.

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 4 above)
Comment (KseniiaMakarova) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: no, this is not a problem, but what about cinemaHall?
my comment was left to line 22)

File: src/main/java/com/dev/movietheatre/Main.java [hunk 8]
```diff
@@ -50,5 +52,16 @@ public static void main(String[] args) throws AuthenticationException {
         User loginedUser = authenticationService.login("Billy@ukr.net", "111");
         System.out.println("Registered user " + registeredUser.toString());
         System.out.println("Logined user " + loginedUser.toString());
+
+        ShoppingCartService shoppingCartService = (ShoppingCartService) injector
+                .getInstance(ShoppingCartService.class);
+        User registeredUser1 = authenticationService.register("Sam@ukr.net", "000");
+        shoppingCartService.addSession(movieSession,registeredUser1);
+        ShoppingCart shoopingCartbyUser1 = shoppingCartService.getByUser(registeredUser1);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]:
```suggestion
        ShoppingCart shoppingCartByUser1 = shoppingCartService.getByUser(registeredUser1);
```
please be careful with spelling and camelCase

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 4 above)
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java]: what about FetchType.LAZY for CinemaHall?

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    @Inject
+    private TicketDao ticketDao;
+
+    @Override
+    public void addSession(MovieSession movieSession, User user) {
+        Ticket ticket = new Ticket();
+        ticket.setMovieSession(movieSession);
+        ticket.setUser(user);
+        ticketDao.add(ticket);
+        ShoppingCart shoppingCartByUser = shoppingCartDao.getByUser(user);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: we have the opportunity to pass to this method the user who is not saved to the DB (or who does not have the shopping cart - admin for example)

if so, your method will save the ticket to the DB anyway, even if on the next line there will be no cart for the user

let's fix this

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/movietheatre/Main.java (hunk 8 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: Sorry, i will be more careful.

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 4 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: i'll delete it also. I've not understanded that comment was about line 22, at first.

Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: deleted, done

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java (hunk 9 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: i'm not sure my simple fix is what you required. Maybe your required more complicated fix.

File: src/main/java/com/dev/movietheatre/Main.java [hunk 10]
```diff
@@ -58,10 +58,10 @@ public static void main(String[] args) throws AuthenticationException {
         User registeredUser1 = authenticationService.register("Sam@ukr.net", "000");
         shoppingCartService.addSession(movieSession,registeredUser1);
         ShoppingCart shoopingCartbyUser1 = shoppingCartService.getByUser(registeredUser1);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]: please, still change the name of this variable

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -22,10 +22,12 @@ public void addSession(MovieSession movieSession, User user) {
         Ticket ticket = new Ticket();
         ticket.setMovieSession(movieSession);
         ticket.setUser(user);
-        ticketDao.add(ticket);
-        ShoppingCart shoppingCartByUser = shoppingCartDao.getByUser(user);
-        shoppingCartByUser.getTickets().add(ticket);
-        shoppingCartDao.update(shoppingCartByUser);
+        if (shoppingCartDao.getByUser(user) != null) {
+            ticketDao.add(ticket);
+            ShoppingCart shoppingCartByUser = shoppingCartDao.getByUser(user);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: let's remove `if()` and just switch lines 26 and 27

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java (hunk 11 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/ShoppingCartServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/Main.java (hunk 10 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: renamed, done

File: src/main/java/com/dev/movietheatre/Main.java [hunk 12]
```diff
@@ -55,12 +55,14 @@ public static void main(String[] args) throws AuthenticationException {
...
                 .getInstance(ShoppingCartService.class);
-        User registeredUser1 = authenticationService.register("Sam@ukr.net", "000");
-        shoppingCartService.addSession(movieSession,registeredUser1);
-        ShoppingCart shoopingCartbyUser1 = shoppingCartService.getByUser(registeredUser1);
-        System.out.println("____There is Sam's shopping cart_____ " + shoopingCartbyUser1);
-        shoppingCartService.clear(shoppingCartService.getByUser(registeredUser1));
-        ShoppingCart userWithClearedShoppingCart = shoppingCartService.getByUser(registeredUser1);
+        User newRegisteredUser = authenticationService.register("Sam@ukr.net", "000");
+        shoppingCartService.addSession(movieSession,newRegisteredUser);
+        ShoppingCart shoopingCartByNewRegisteredUser = shoppingCartService
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]: shoOping -> shoPPing

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**